File: Python/1_07.py
# Flatten a nested list structure

nestedList = [1, 2, 3, 4, [5, 6, [66, 666], 12], 7, 8, 9]
nestedList2 = ['a', ['b', ['c', 'd'], 'e']]
t = [0, 1, [2, 3], 4, 5, [6, 7]]

def recFlatten(nestedListParam):
    flatList = []


    for i in range(len(nestedListParam)):
        if type(nestedListParam[i]) == list:
            flatList.extend(recFlatten(nestedListParam[i]))
        else:
            flatList.append(nestedListParam[i])
    
    return flatList

print(recFlatten(nestedList))
print(recFlatten(nestedList2))
print(recFlatten(t))

File: ocaml/1_07.ml
(* Flatten a nested list structure *)

let flatten l = l

let rec test l = match l with
  | [] -> []
  | [hd] -> [hd]
  | hd :: hd' :: tl -> 
    if hd' = hd then test (hd':: tl)
    else hd :: test (hd' :: tl)

let () = 
  List.iter (fun e -> Printf.printf " %d" e) (test [1;1;2;6;7;4;45;879;879;5;5;4;5;2;2;2;4]);
  print_endline "";